### DISCUSSION TITLE
Incremental Static Regeneration

### DISCUSSION THREAD

**Discussion by timmske** [vercel/next.js#24022] [Help]
Hello everybody,

I have a question about the Incremental Static Regeneration. As far as I know the revalidate value within the getStaticProps() function tells NextJS the amount of time it should rebuild the pages.

My question about that is, will this happen for every user/request after the set amount of time, or centralized, starting with the first user/request hitting the page?

#############################################

**For example:**

Revalidate value in the getStaticProps() function: **60 seconds**

**_User A_** hits the page and receives a cached version. After **60 seconds** NextJS rebuilds the pages for him and serves the fresh content.

**_User B_** hits the page shortly after User A, receives a cached version and after 60 seconds he also receives an updated version.

#############################################

My concern is that every individual request starts it's **own** 60-second interval to rebuild.

I'm quite sure that's not the case, but as NextJS is new to me, I wanted to get this straight before messing things up.

I would be very thankful If somebody could volunteer to give a quick response. It would help me a lot!

Thanks in advance to everybody!

**Comment by nghiepdev** [comment]
I am also interested in cache time on ISR.

**For example:**
Revalidate value in the getStaticProps() function: **60** seconds

**User A** hist the page and receives a cached version. In **60 seconds** Next.js rebuild the page.
About 2000 years later **User B** revisit again this page => OMG, He just only will receive the cached of the previous 2000 years.
For the next request, a new version will be returned.

**Comment by michrome** [comment]
Hi @timmske! Don’t think of it as per _user_; think about it as per _page_. Using your 60s example:

* The first time the page is requested, it is generated (let’s call it v1), cached then served.
* Any requests for that page made in the next 60 seconds will be served the same cached v1 copy.
* The first request after 60 seconds will also be served the same cached v1 copy. Although v1 was served, it is considered stale, so a new v2 is generated and cached (but not served).
* The next request will be served the cached v2 copy.

**Reply by timmske** [reply]
Hey,
first of all thank you very much for your explanation!

Does the comparison/check for a newer version cause heavy server load when dealing with thousands of requests? 

Also, is there another way to serve static content and to "manually" rebuild all pages after a given amount of time, independently of the requests? I read about the official Vercel host being capable of that, but is there an option when deploying your project to a different provider?

**Reply by mbrochh** [reply]
> Hi @timmske! Don’t think of it as per _user_; think about it as per _page_. Using your 60s example:
> 
>     * The first time the page is requested, it is generated (let’s call it v1), cached then served.
> 
>     * Any requests for that page made in the next 60 seconds will be served the same cached v1 copy.
> 
>     * The first request after 60 seconds will also be served the same cached v1 copy. Although v1 was served, it is considered stale, so a new v2 is generated and cached (but not served).
> 
>     * The next request will be served the cached v2 copy.

Doesn't this feature pose quite a costly risk for sites that have hundreds of thousands of pages? If the Google Bot comes along and crawls all the pages, wouldn't it cause re-renders of basically all pages because all the content will be stale?

**Comment by michrome** [comment]
> first of all thank you very much for your explanation!

No problem 😀

> Does the comparison/check for a newer version cause heavy server load when dealing with thousands of requests?

Categorically: **no**.

> Also, is there another way to serve static content and to "manually" rebuild all pages after a given amount of time, independently of the requests? I read about the official Vercel host being capable of that, but is there an option when deploying your project to a different provider?

On demand is coming in the future: “[We'll be exploring on-demand (triggered via API route) revalidation instead of interval-based in the future.](https://github.com/vercel/next.js/discussions/11552#discussioncomment-2655)”

> Doesn't this feature pose quite a costly risk for sites that have hundreds of thousands of pages? If the Google Bot comes along and crawls all the pages, wouldn't it cause re-renders of basically all pages because all the content will be stale?

I don’t think so. For pages that didn’t change often, you’d set their realisation period to a higher number (regeneration only happens when the content is stale),or not use Incremental Static Regeneration for those pages.
